File: spike/remotion/tokens.ts
/**
 * 스타일 자산. AGENTS.md §9.
 *
 * ★ 이 파일의 값은 **`spike/public/frames/` 를 눈으로 재서** 채운다.
 *   추측해서 채우지 않는다. 전작이 기본값을 디자인 시안 색(#E8C33F 노란 박스)으로
 *   박아 놓고 "스타일 학습"이라고 부른 게 §0-4 실패 원인이다.
 *
 * 아래 값은 유튜브 숏츠 1편을 화면으로만 본 **미측정 추정치**다.
 * `pnpm spike:frames` 로 프레임을 뽑고 실제로 재서 전부 덮어쓴 뒤 MEASURED 를 true 로 바꾼다.
 */

export const MEASURED = false as boolean;

/**
 * 잠정 실측 (2026-09-23).
 *
 * 출처: 유튜브 숏츠 `EDpBGkaNJmU` 1편, 브라우저 canvas 캡처 10프레임, 720x1280.
 * 아래 값은 여기서 재서 1080x1920 으로 환산했다.
 *
 * ★ 아직 MEASURED=false 인 이유:
 *   - 1편만 쟀다. §9 절차는 5편을 요구한다
 *   - 유튜브 720p 재인코딩본이다. 인스타 릴스 원본과 자막이 다를 수 있다
 *     (이 영상은 auto-dubbed 버전이라 영문 보조자막이 붙어 있다)
 *   - reference/final.mp4 가 들어오면 다시 재고 그때 true 로 바꾼다
 *
 * 확정된 판정:
 *   - 자막 크기는 **고정**이다. fit-to-width 가 아니다.
 *     글자 수 9~12자에서 폭 비율은 0.61~0.82 로 변하는데 글자 높이는 47~49px 로 일정했다.
 *     → 폭 기준 자동 스케일을 구현하지 않는다.
 *   - 강조색은 쓰지 않는다. 본문은 전부 흰색이었다.
 *   - 보조 문구(영문)는 **노란색**이 맞다.
 */

export const FRAME = {
  width: 1080,
  height: 1920,
  fps: 30,
} as const;

/**
 * 자막 본문.
 * 품질 게이트 G4: 글자 높이 >= 프레임 높이의 4.5% (1920 기준 86px).
 * 전작은 fontSize 56 (2.9%) 이었고 그래서 자동 자막처럼 보였다.
 */
export const CAPTION = {
  fontFamily: 'Pretendard Variable, Pretendard, sans-serif',
  /**
   * px. 1920 높이 기준.
   * 원본 실측: 글자 높이 46px @1280 = 3.59% → 1920 환산 69px.
   * 프로브 측정: fontSize 72 가 글자 높이 64px 을 냈다 (글자높이 ≈ fontSize x 0.889).
   *   → 69 / 0.889 ≈ 78
   * 바꾸면 `pnpm spike:measure out/caption-probe.png` 로 다시 확인한다.
   */
  fontSize: 78,
  fontWeight: 800,
  color: '#FFFFFF',
  /** 검은 외곽선. strokeWidth 는 **바깥으로 나간** 두께다 (Caption.tsx 가 2배로 넘긴다). */
  strokeColor: '#000000',
  /** 실측: 4~5px @720 → 1080 환산 6~7.5px */
  strokeWidth: 7,
  /** 화면 폭 대비 최대 사용 비율 */
  maxWidthRatio: 0.88,
  /**
   * 화면 아래에서 띄우는 비율. **본문 아래끝** 기준이다 (보조 문구는 이 아래에 매달린다).
   * 실측: 4개 프레임에서 0.2344~0.2352 로 거의 흔들리지 않았다. 가장 신뢰도 높은 값.
   */
  bottomRatio: 0.235,
  lineHeight: 1.18,
  /**
   * 줄상자(line box) 아래쪽 여백 보정. fontSize 대비 비율.
   *
   * bottomRatio 는 **글자 아래끝**을 뜻하는데, CSS 는 글자가 아니라 줄상자를 배치한다.
   * 줄상자 아래에는 leading 과 디센더 공간이 남아서 글자가 그만큼 위로 뜬다.
   * 프로브 측정: 목표 0.2350 에 0.2406 이 나왔다 = 11px @1920 떠 있었다. 11/72 ≈ 0.153.
   *
   * 폰트 · lineHeight · fontSize 를 바꾸면 다시 잰다. measure.mjs 가 바로 잡아낸다.
   */
  baselineNudgeRatio: 0.153,
  /** 한 덩어리 최대 글자 수. 실측: 12자까지 한 줄로 갔다. 품질 게이트 G5 */
  maxChars: 13,
  maxLines: 2,
  /** 등장 애니메이션 */
  popInMs: 100,
  popInScaleFrom: 0.86,
} as const;

/**
 * 강조 단어. 색만 바꾼다. 크기를 바꾸면 줄바꿈이 흔들린다.
 * ★ 실측 영상에서는 강조를 **쓰지 않았다**. 본문은 전부 흰색.
 *   기능은 남겨두되 기본으로 쓰지 않는다. 쓸 근거가 생기면 그때 색을 정한다.
 */
export const CAPTION_EMPHASIS = {
  color: '#FFE04D',
} as const;

/** 보조 문구(영문 등). 본문 바로 아래. */
export const CAPTION_SECONDARY = {
  /** 실측: 본문 약 48px 대비 보조 약 22px @720 → 0.46 */
  scale: 0.48,
  color: '#FFE04D',
  italic: true,
  strokeWidth: 5,
  /**
   * 본문 아래끝과 보조 문구 윗끝 사이 간격 (본문 fontSize 대비).
   * 원본 실측: 본문 아래끝 0.235H, 보조 아래끝 0.204H → 둘의 차이 59.5px @1920.
   * 프로브 측정: fontSize 72 · gapRatio 0.34 에서 차이가 75px 이었다.
   *   75 = 72 x (0.34 + C)  ->  C = 0.701 (보조 줄상자 + 글자 높이 몫)
   *   fontSize 78 에서 59.5 를 맞추려면 78 x (g + 0.701) = 59.5  ->  g ~= 0.062
   */
  gapRatio: 0.062,
} as const;

/** 훅 타이틀. 품질 게이트 G8: 0~1.5초. */
export const HOOK = {
  fontSize: 110,
  fontWeight: 900,
  color: '#FFFFFF',
  strokeColor: '#000000',
  strokeWidth: 10,
  /** 화면 위에서 띄우는 비율 */
  topRatio: 0.14,
  durationSec: 1.2,
  zoomFrom: 1.0,
  zoomTo: 1.06,
} as const;

/**
 * 리프레이밍 목표.
 * 품질 게이트 G1: 인물 bbox 높이 >= 프레임 높이의 55% 인 구간이 80% 이상.
 * 목표치는 그보다 넉넉하게 잡는다.
 */
export const REFRAME = {
  /** 인물 bbox 높이가 프레임 높이의 이만큼을 채우도록 */
  targetSubjectHeightRatio: 0.72,
  /** 저역통과 스무딩 시간(초). 이보다 짧으면 프레임이 떨린다 (G3) */
  smoothingSec: 0.4,
  /** 피사체 주변 여백 */
  padding: 0.08,
} as const;

export const AUDIO = {
  bgmGainDb: -22,
  bgmDuckDb: -6,
  sfxGainDb: -8,
  /** 품질 게이트 G12 */
  targetLufs: -14,
} as const;

/** 값을 실측 전에 렌더하면 경고. 조용히 지나가지 않는다. */
export function warnIfUnmeasured(): void {
  if (!MEASURED) {
    console.warn(
      '\n[tokens] 공개본 1편으로만 맞춘 잠정값입니다. reference/final.mp4 로 5편 재측정 후 MEASURED=true 로 바꾸세요.\n' +
        '  AGENTS.md §9 / docs/stage-0.spec.md 작업순서 3~4\n',
    );
  }
}

File: Madi/Templates/SuhyunShortV1/Tokens.swift
import Foundation
import CoreGraphics

/// 스타일 자산. AGENTS.md §9 · docs/style-authoring.md.
///
/// ★ 이 파일의 값은 **`reference/` 프레임을 픽셀로 재서** 채운다. 추측해서 채우지 않는다.
///   전작이 기본값을 디자인 시안 색(#E8C33F 노란 박스, 56px)으로 박아 놓고
///   "스타일 학습" 이라고 부른 게 §0-4 실패 원인이다.
///
/// ★ 값은 전부 **프레임 높이 대비 비율**이다. 포인트 크기가 아니다.
///   "글자 높이 ≠ 폰트 크기" 이므로 폰트 크기는 `CaptionLayer` 가 CoreText 메트릭에서 역산한다
///   (docs/stage-0.spec.md 알려진 함정). 여기에 포인트 값을 적으면 폰트를 바꿀 때마다 전부 틀어진다.
///
/// 측정 출처: `reference/yt_15s.png` (720x1280).
/// 이 프레임은 `tools/measure.mjs` 의 세 목표치를 모두 통과하는 골든 프레임이다
/// (본문높이 3.59% · 하단여백 0.2352 · 보조하단 0.2039).
public enum SuhyunShortV1 {

    /// 값이 `docs/style-authoring.md §1` 의 5편 절차를 통과했는가.
    ///
    /// 아직 false 인 이유: 자막 실측은 유튜브 숏츠 1편(`EDpBGkaNJmU`)에서만 했다.
    /// 레이아웃 어휘는 5편을 봤지만(`docs/findings/2026-09-23-layout-survey.md`)
    /// 자막 치수는 아니다. 인스타 릴스 원본이 들어오면 5편으로 다시 재고 true 로 바꾼다.
    public static let measured = false

    public enum Frame {
        public static let width = 1080
        public static let height = 1920
        public static let fps = 30
    }

    // MARK: - 자막 본문

    public enum Caption {
        /// `wght` 가변 축.
        ///
        /// 실측 근거: `yt_15s.png` 에서 세로 획의 흰 폭이 5px, 글자 높이가 46px 이었다
        /// (획/글자높이 = 0.109). 이 비율이 나오는 웨이트를 `madi-spike stem` 으로 골랐다.
        /// 값을 바꾸면 반드시 다시 재고 이 주석을 고친다.
        public static let weight: CGFloat = 800

        /// **그려진 흰 글자의 높이** ÷ 프레임 높이.
        /// 실측 46px / 1280 = 0.0359. 신뢰도 높음 — 3편에서 같은 값.
        /// (`tools/measure.mjs` 가 흰 픽셀 밴드로 재는 값과 같은 정의다)
        public static let inkHeightRatio: Double = 0.0359

        /// 흰 글자 **아래끝**에서 화면 아래까지 ÷ 프레임 높이.
        /// 실측 0.2344~0.2352 (4프레임). 가장 흔들림이 적은 값이라 신뢰도 높음.
        /// ★ 블록 전체가 아니라 **본문 마지막 줄의 아래끝** 기준이다.
        ///   블록 기준으로 잡으면 보조 문구가 있을 때 본문이 밀려 올라간다 (archive/README.md).
        public static let inkBottomRatio: Double = 0.2352

        /// 검은 외곽선이 글자 **바깥으로** 나간 두께 ÷ 프레임 높이.
        /// 실측: 글자 위쪽 검은 띠 4px @1280 (= 6px @1920). 4/1280 = 0.003125.
        ///
        /// ★ 원본은 외곽선이 글자 안쪽을 **깎지 않는다.** 흰 픽셀 바로 위가 검은 4px 이고
        ///   그 안쪽은 온전히 흰색이다. 즉 "stroke 먼저, fill 나중" 순서로 그려졌다.
        ///   `NSAttributedString.strokeWidth` 를 음수로 주면 fill 먼저 stroke 나중이라
        ///   글자 안쪽이 획 절반만큼 깎인다 — 원본과 다르다. `CaptionLayer` 는 2패스로 그린다.
        public static let strokeOuterRatio: Double = 0.003125

        /// 화면 폭 대비 상한. **목표가 아니라 clamp 다** (docs/style-authoring.md).
        /// 참고: 실측에서 11자가 폭의 0.657 을 썼다. 12자여도 0.72 근처라 이 상한은 거의 안 걸린다.
        public static let maxWidthRatio: Double = 0.90

        /// 한 덩어리 최대 글자 수. 실측: 12자까지 한 줄로 갔다. 품질 게이트 G5.
        public static let maxChars = 13
        public static let maxLines = 2

        /// 2줄일 때 줄 간격 ÷ 글자 높이.
        /// ⚠ **미측정.** `reference/` 34장에 2줄 자막 프레임이 없었다.
        ///   폰트의 기본 행간(ascent+descent+leading)을 쓰면 한글에서 너무 벌어져서
        ///   1.35 로 뒀다. 2줄 자막이 나오는 프레임을 확보하면 재서 고친다.
        public static let lineGapRatio: Double = 1.35

        /// 등장 애니메이션. 픽셀로 못 재므로 연속 프레임을 세서 잡는다
        /// (docs/style-authoring.md §2). ⚠ 미측정 — 눈에 띄게 다르지만 않으면 된다.
        public static let popInSec: Double = 0.1
        public static let popInScaleFrom: CGFloat = 0.86
    }

    /// 본문 색. 실측: rgb(255,255,255) 순백.
    public static let captionFill = RGBA(1, 1, 1, 1)
    /// 외곽선 색. 실측: rgb(0,0,0).
    public static let captionStroke = RGBA(0, 0, 0, 1)

    /// 강조 단어.
    /// ★ 실측 영상에서는 강조를 **쓰지 않았다.** 본문은 전부 흰색이다.
    ///   기능은 남겨 두되 기본으로 쓰지 않는다. 쓸 근거가 생기면 그때 색을 정한다.
    public static let captionEmphasis = RGBA(1, 0.878, 0.302, 1) // #FFE04D

    // MARK: - 보조 문구 (영문 등)

    public enum Secondary {
        /// 본문 폰트 크기 대비 비율.
        /// 실측: 라틴 어센더 높이 16px, 베이스라인 y=1020 @1280.
        /// 본문 크기를 역산한 뒤 어센더 16px 이 나오는 비율을 `madi-spike metrics` 로 골랐다.
        public static let scale: CGFloat = 0.40

        /// 웨이트. 실측 이미지에서 획이 가늘고 균일 — Regular~Medium.
        public static let weight: CGFloat = 500

        /// 보조 문구 **베이스라인**에서 화면 아래까지 ÷ 프레임 높이.
        ///
        /// ★ 본문과 달리 ink 아래끝이 아니라 베이스라인으로 잡는다. 라틴 문자는 `y` · `g` 의
        ///   디센더가 얼마나 내려오는지가 글자마다 달라서 ink 아래끝이 문구에 따라 흔들린다.
        ///   베이스라인은 안 흔들린다.
        /// 실측: 행별 노란 픽셀 수가 y=1020 에서 46 → 6 으로 끊긴다 = 베이스라인 1020 @1280.
        ///   (1280 - 1020 - 1) / 1280 = 0.2023
        /// `tools/measure.mjs` 는 임계값 때문에 이 자막의 아래끝을 0.2039 로 읽는다. 그게 목표치다.
        public static let baselineBottomRatio: Double = 0.2023

        /// ★ 이탤릭이 **아니다.** 6배 확대해서 확인했다 — 글자가 곧게 서 있다.
        ///   (archive 의 Remotion 토큰은 `italic: true` 였다. 재보니 틀렸다)
        public static let italic = false

        /// ★ 외곽선이 **없다.** 노란 글자에 검은 테두리가 없고 부드러운 그림자만 있다.
        public static let hasStroke = false
    }

    /// 보조 문구 색. 실측: rgb(254,227,116) = #FEE374.
    public static let secondaryFill = RGBA(254.0 / 255, 227.0 / 255, 116.0 / 255, 1)

    // MARK: - 그림자

    /// 자막 아래에 깔리는 부드러운 그림자.
    ///
    /// 실측 근거: 본문 글자 위쪽 검은 띠는 4px 에서 배경으로 **딱 끊기는데**(929~932),
    /// 아래쪽은 5px 뒤 2px 에 걸쳐 배경으로 번진다(979~985, 루마 0 → 35 → 64 → 배경 81).
    /// 위아래 비대칭 + 아래쪽 번짐 = 아래로 조금 내린 흐린 그림자다.
    /// 보조 문구에서는 더 뚜렷하다 (외곽선이 없는데도 오른쪽 아래가 어둡다).
    ///
    /// ⚠ 신뢰도 낮음. 유튜브 720p 재인코딩본이라 압축 번짐과 구분이 어렵다.
    ///   `enabled` 를 false 로 두면 외곽선만 그린다.
    public enum Shadow {
        public static let enabled = true
        /// 프레임 높이 대비. 실측 1px @1280.
        public static let offsetRatio = CGSize(width: 1.0 / 1280, height: -1.0 / 1280)
        /// 프레임 높이 대비. 실측 번짐 폭 2px @1280.
        public static let blurRatio: Double = 2.0 / 1280
        public static let color = RGBA(0, 0, 0, 0.55)
    }

    // MARK: - 훅 · 리프레이밍 · 오디오 (1~2단계)

    /// ⚠ 전부 **미측정**. 0단계 범위 밖이라 재지 않았다.
    ///   1단계(리프레이밍)·2단계(템플릿 완성)에서 프레임을 재서 채운다.
    ///   지금 이 값으로 렌더하면 원본과 다르게 나온다 — 그래서 0단계 스파이크는 쓰지 않는다.
    public enum Reframe {
        /// 품질 게이트 G1 은 0.55 를 하한으로 본다. 목표는 그보다 넉넉히.
        public static let targetSubjectHeightRatio: Double = 0.72
        /// 이보다 짧게 스무딩하면 프레임이 떨린다 (G3).
        public static let smoothingSec: Double = 0.4
        public static let padding: Double = 0.08
    }
}

/// 0..1 색. `CGColor` 로 바꿔 쓴다.
public struct RGBA: Hashable, Sendable {
    public var r: CGFloat, g: CGFloat, b: CGFloat, a: CGFloat
    public init(_ r: CGFloat, _ g: CGFloat, _ b: CGFloat, _ a: CGFloat) {
        self.r = r; self.g = g; self.b = b; self.a = a
    }
    public var cgColor: CGColor {
        CGColor(colorSpace: CGColorSpace(name: CGColorSpace.sRGB)!, components: [r, g, b, a])!
    }
}
